### DISCUSSION TITLE
Simulating Rollbacks Locally

### DISCUSSION THREAD

**Discussion by jeremyhcho** [foxssake/netfox#210] [Q&A] [answered]
Hi! I'm in the process of considering netcoding options in Godot so I'm putting together a small proof of concept for myself to implement and test things.

I was wondering if there was an easy way to essentially create an environment where rollbacks are constantly triggered on clients to test the netcoding in various situations. What would be the best way to do this with Netfox?

**Comment by elementbound** [comment, marked as answer]
Hi @jeremyhcho! Rollback will happen, as long as there's active RollbackSynchronizers and latency between server and clients. The higher the latency, the older the input received on the server, and the more ticks it needs to resimulate. Similarly on clients - the higher the latency, the older the last received state, so more ticks need to be resimulated.

For local testing on Linux, there's a [script](https://github.com/foxssake/netfox/blob/main/sh/latency.sh) included that you can use to simulate latency artifically. For Windows, I know there's [clumsy](https://jagt.github.io/clumsy/), although I haven't used it myself.

You can check the number of rollback ticks simulated in Godot's profiler:

![image](https://github.com/foxssake/netfox/assets/4706096/31198980-f351-4964-8c59-5e906c477c14)

Out of curiosity, what are the alternatives you're checking? I only know of [Snopek's rollback addon](https://gitlab.com/snopek-games/godot-rollback-netcode/).

**Reply by jeremyhcho** [reply]
Gotcha, thanks for the quick response!

And yeah at the moment I'm currently just looking at Snopek's as an alternative.
